### ISSUE TITLE
Working code might break under `--no-check` & will still be broken after removing `--no-check`

### ISSUE THREAD

**Issue by WilcoKruijer** [denoland/deno#10684]
Not exactly sure what the issue is, but it can be reliably reproduced. This was an issue in Deno 1.9, which still exists in 1.10.2.

## Reproduction

Create a file `a.ts`:

```typescript
import { deserializeFeed } from "https://deno.land/x/rss@0.3.3/mod.ts";

const response = await fetch("https://hnrss.org/newest");
const xml = await response.text();
const { feed } = await deserializeFeed(xml, { outputJsonFeed: true });

console.dir(feed);

```

Run using:
`deno run --allow-net a.ts`

This will work no problem.

Now we rerun using `--no-check`:

This gives us the following error:
```
$ deno run --allow-net --no-check a.ts

error: Uncaught SyntaxError: The requested module './types/mod.ts' does not provide an export named 'JsonFeed'
        JsonFeed,
        ~~~~~~~~
    at <anonymous> (https://deno.land/x/rss@0.3.3/src/deserializer.ts:7:2)
```

Removing `--no-check` will still give us the error:
```
$ deno run --allow-net a.ts

error: Uncaught SyntaxError: The requested module './types/mod.ts' does not provide an export named 'JsonFeed'
        JsonFeed,
        ~~~~~~~~
    at <anonymous> (https://deno.land/x/rss@0.3.3/src/deserializer.ts:7:2)
```

Even using `--reload` will not help.
```
$ deno run --allow-net --reload a.ts

error: Uncaught SyntaxError: The requested module './types/mod.ts' does not provide an export named 'JsonFeed'
        JsonFeed,
        ~~~~~~~~
    at <anonymous> (https://deno.land/x/rss@0.3.3/src/deserializer.ts:7:2)
```

Now we rename the file:
`mv a.ts b.ts`

And rerun the original command which will work just fine:
`deno run --allow-net b.ts`

**Comment by kitsonk** [comment]
Hmmm. `--reload` not working is undesired. The other part of this is an swc emit error.

**Comment by kitsonk** [comment]
Cc/ @kdy1

**Comment by paul90** [comment]
I've got the same problem trying to use this with deploy. Works fine using `deployctl` locally, without `--no-check`, but deploy gives the error shown above. While renaming the file works locally with `deployctl`, it doesn't help with deploy (which I guess is much as expected).

**Comment by kitsonk** [comment]
Actually, I originally looked at this on my phone and didn't realise the problem.  It isn't an swc issue.  It is that the dependency, sadly, is not "transpile safe" and the way they are being exported does not trigger the "isolated modules" setting in the Deno CLI to give you an indication that it isn't working.

The author of the package should be a) using `.d.ts` files for type only files and b) use `export type` and `import type` instead of just `export` and `import`.

This is because when you are transpiling only (Deploy and `--no-check`), the transpiler has no idea that what you are trying to import/export is just a type that gets erased at runtime and so when it tries to do the named export (https://deno.land/x/rss@0.3.3/mod.ts#L8) it has no idea that that is just a type that got erased.  You can read more about type only imports and exports here: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export.  Again, we have the built in TypeScript settings trying to catch this, but the package author has found an interesting way to get around the `--isolatedModules` setting unintentionally, by doing `export *` in the `./types/mod.ts`.

I would open an issue with the package author.

As far as `--reload` not working, that still needs investigation.

**Comment by paul90** [comment]
Thank you. I'll raise an issue with the package author.

**Comment by nayeemrmn** [comment]
This is an swc bug. The reason it doesn't trigger an `--isolatedModules` error is because transpilers are supposed to handle these cases just fine. And a type-only module is valid TypeScript that should emit an empty JavaScript module, it doesn't or shouldn't need a `d.ts` extension.

Smaller repro:
```ts
// temp2.ts
export type Foo = null;
```
```ts
// temp.ts
import { Foo } from "./temp2.ts";

const a: Foo = null;
console.log(a);
const b = { Foo: 1 };
console.log(b.Foo); // This wrongly counts as a value usage of the imported `Foo`, causing the import to be preserved.
```
Actual transpilation of `temp.ts`:
```js
import { Foo } from "./temp2.ts";
const a = null;
console.log(a);
const b = {
    Foo: 1
};
console.log(b.Foo);
```
Expected:
```js
const a = null;
console.log(a);
const b = {
    Foo: 1
};
console.log(b.Foo);
```
@kdy1 This is the same issue as #10462, which wasn't fixed "robustly".

**Comment by kdy1** [comment]
I tried this, but https://github.com/swc-project/swc/blob/15e7da2ea8cc942369c7a9c059bfbf2ff37662a0/tests/fixture/deno-10684/case1/output/entry.ts does not contain `import { Foo } from "./temp2.ts";`
I'll try something smaller.

**Comment by kdy1** [comment]
Oh, I was able to reproduce it using `typescript::strip`. I'll fix this asap.

**Comment by kdy1** [comment]
Fixed on upstream (`swc_ecma_transforms_typescript`)

**Comment by nayeemrmn** [comment]
Fixed by #11626. 

There seemed to also be some unexpected behaviour with `--reload` but I'm not sure it can manifest without this kind of emit bug. We should probably forget about it unless someone reproduces it again.
